{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e8dced2d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MSLLNSTKWETEGALVALERGKLPSTDTY\n",
      "MR\n",
      "MGAMKGSDCDR\n",
      "MKGSDCDR\n",
      "MCTDSSLCT\n",
      "MEALGQAYELLRVGSCRVGGLSLFADCRVFGGSEDSSDKFCELSSVRMGRFKVA\n",
      "MGRFKVA\n",
      "MPRSQVPRTLAMFNPHESMRTRAFEI\n",
      "MFNPHESMRTRAFEI\n",
      "MRTRAFEI\n",
      "MGNGRCSSCPGAWQVTIDGHLLA\n",
      "MSSAYTQCAPIRVSAPSEVRSTKIPRSGPQIR\n",
      "MNYCGLVLVG\n",
      "MLTGASSASVVGSPYLRATTGYLSAPYLARCRDSNRCTLSICRRHEGQ\n",
      "MSAPRLEWTVASQIPGLDFKCTGTHRFVRVKHC\n",
      "MINLCGPPRLWVLRRQYRC\n",
      "MVQASRCPSMVTCHAPGQLEHLPFPILWSLVGSCQRLG\n",
      "MVTCHAPGQLEHLPFPILWSLVGSCQRLG\n",
      "MFQAFLIV\n",
      "MPGS\n",
      "MRTLESSQNLSLESSDPPNTRQSANKLSPPTLQEPTRSSS\n",
      "MRNYERLTNTLPHMS\n",
      "MS\n",
      "MQKT\n",
      "MHGYA\n"
     ]
    }
   ],
   "source": [
    "aminos = {'AAA': 'K', 'AAC': 'N', 'AAG': 'K', 'AAT': 'N', 'ACA': 'T', 'ACC': 'T', 'ACG': 'T', 'ACT': 'T', 'AGA': 'R', 'AGC': 'S', 'AGG': 'R', 'AGT': 'S', 'ATA': 'I', 'ATC': 'I', 'ATG': 'M', 'ATT': 'I', 'CAA': 'Q', 'CAC': 'H', 'CAG': 'Q', 'CAT': 'H', 'CCA': 'P', 'CCC': 'P', 'CCG': 'P', 'CCT': 'P', 'CGA': 'R', 'CGC': 'R', 'CGG': 'R', 'CGT': 'R', 'CTA': 'L', 'CTC': 'L', 'CTG': 'L', 'CTT': 'L', 'GAA': 'E', 'GAC': 'D', 'GAG': 'E', 'GAT': 'D', 'GCA': 'A', 'GCC': 'A', 'GCG': 'A', 'GCT': 'A', 'GGA': 'G', 'GGC': 'G', 'GGG': 'G', 'GGT': 'G', 'GTA': 'V', 'GTC': 'V', 'GTG': 'V', 'GTT': 'V', 'TAA': '*', 'TAC': 'Y', 'TAG': '*', 'TAT': 'Y', 'TCA': 'S', 'TCC': 'S', 'TCG': 'S', 'TCT': 'S', 'TGA': '*', 'TGC': 'C', 'TGG': 'W', 'TGT': 'C', 'TTA': 'L', 'TTC': 'F', 'TTG': 'L', 'TTT': 'F'}\n",
    "\n",
    "file = 'rosalind_orf.txt'\n",
    "\n",
    "with open(file, 'r') as dna:  #open fasta file\n",
    "    data = dna.readlines()\n",
    "\n",
    "seq = ''\n",
    "\n",
    "for row in data[1:]:      #iterate through file after header and strip newline characters\n",
    "    fasta = (row.strip('\\n'))\n",
    "    seq = seq + fasta     #join to empty list\n",
    "\n",
    "dna.close()  #close file\n",
    "\n",
    "orf1 = []\n",
    "orf1_proteins = ''\n",
    "\n",
    "\n",
    "for aa in range(0,len(seq),3):   #iterate through dna seq and append every three letters to new list as codons\n",
    "    orf1.append(seq[aa:aa+3])\n",
    "\n",
    "for i in orf1:     #iterate through list of codons and store its amino acid symbol to empty string, making protein sequence\n",
    "    if i in aminos.keys():\n",
    "        orf1_proteins += aminos[i]\n",
    "\n",
    "####### coding sequence ##########\n",
    "coding_seq = []\n",
    "\n",
    "for i in range(len(orf1_proteins)):\n",
    "    for j in range(len(orf1_proteins)):\n",
    "        if orf1_proteins[i] == 'M':\n",
    "            if orf1_proteins[j] == '*':\n",
    "                coding_seq.append(orf1_proteins[i:j:])\n",
    "\n",
    "######### coding sequence ##########\n",
    "\n",
    "orf2 = []\n",
    "orf2_proteins = ''\n",
    "\n",
    "\n",
    "for aa2 in range(1,len(seq),3):   #Repeat methodology in next ORF, but begin iterating through dna seq on the second character\n",
    "    orf2.append(seq[aa2:aa2+3])\n",
    "    \n",
    "for codon in orf2:        #remove item in list if it is not a full three letter codon\n",
    "    if len(codon) != 3:\n",
    "        orf2.remove(codon)\n",
    "\n",
    "for i2 in orf2:     #translate to proteins and store in empty string\n",
    "    if i2 in aminos.keys():\n",
    "        orf2_proteins += aminos[i2]\n",
    "\n",
    "for i in range(len(orf2_proteins)):\n",
    "    for j in range(len(orf2_proteins)):\n",
    "        if orf2_proteins[i] == 'M':\n",
    "            if orf2_proteins[j] == '*':\n",
    "                coding_seq.append(orf2_proteins[i:j:])\n",
    "                \n",
    "##### ORF3 #########\n",
    "\n",
    "orf3 = []\n",
    "orf3_proteins = ''\n",
    "\n",
    "\n",
    "for aa3 in range(2,len(seq),3):     #same methodology as above, but again begin iterating one character down.\n",
    "    orf3.append(seq[aa3:aa3+3])\n",
    "    \n",
    "for codon3 in orf3:\n",
    "    if len(codon3) != 3:\n",
    "        orf3.remove(codon3)\n",
    "\n",
    "for i3 in orf3:\n",
    "    if i3 in aminos.keys():\n",
    "        orf3_proteins += aminos[i3]\n",
    "        \n",
    "for i in range(len(orf3_proteins)):\n",
    "    for j in range(len(orf3_proteins)):\n",
    "        if orf3_proteins[i] == 'M':\n",
    "            if orf3_proteins[j] == '*':\n",
    "                coding_seq.append(orf3_proteins[i:j:])\n",
    "                \n",
    "####### REVERSE COMPLEMENT #########\n",
    "\n",
    "rna = ''\n",
    "\n",
    "for nt in seq:\n",
    "    if nt == 'A':\n",
    "        rna += 'T'\n",
    "    \n",
    "    elif nt == 'T':\n",
    "        rna += 'A'\n",
    "        \n",
    "    elif nt == 'G':\n",
    "        rna += 'C'\n",
    "        \n",
    "    elif nt == 'C':\n",
    "        rna += 'G'\n",
    "\n",
    "rev_com = (rna[::-1])\n",
    "\n",
    "                \n",
    "####### ORF4 ##########\n",
    "\n",
    "orf4 = []\n",
    "orf4_proteins = ''\n",
    "\n",
    "\n",
    "for aa in range(0,len(rev_com),3):   #iterate through dna seq and append every three letters to new list as codons\n",
    "    orf4.append(rev_com[aa:aa+3])\n",
    "\n",
    "for i in orf4:     #iterate through list of codons and store its amino acid symbol to empty string, making protein sequence\n",
    "    if i in aminos.keys():\n",
    "        orf4_proteins += aminos[i]\n",
    "\n",
    "for i in range(len(orf4_proteins)):\n",
    "    for j in range(len(orf4_proteins)):\n",
    "        if orf4_proteins[i] == 'M':\n",
    "            if orf4_proteins[j] == '*':\n",
    "                coding_seq.append(orf4_proteins[i:j:])\n",
    "                \n",
    "###### ORF 5 #############\n",
    "\n",
    "orf5 = []\n",
    "orf5_proteins = ''\n",
    "\n",
    "\n",
    "for aa2 in range(1,len(rev_com),3):   #Repeat methodology in next ORF, but begin iterating through dna seq on the second character\n",
    "    orf5.append(rev_com[aa2:aa2+3])\n",
    "    \n",
    "for codon in orf5:        #remove item in list if it is not a full three letter codon\n",
    "    if len(codon) != 3:\n",
    "        orf5.remove(codon)\n",
    "\n",
    "for i2 in orf5:     #translate to proteins and store in empty string\n",
    "    if i2 in aminos.keys():\n",
    "        orf5_proteins += aminos[i2]\n",
    "\n",
    "for i in range(len(orf5_proteins)):\n",
    "    for j in range(len(orf5_proteins)):\n",
    "        if orf5_proteins[i] == 'M':\n",
    "            if orf5_proteins[j] == '*':\n",
    "                coding_seq.append(orf5_proteins[i:j:])\n",
    "\n",
    "######## ORF 6 ###########\n",
    "\n",
    "orf6 = []\n",
    "orf6_proteins = ''\n",
    "\n",
    "\n",
    "for aa3 in range(2,len(rev_com),3):     #same methodology as above, but again begin iterating one character down.\n",
    "    orf6.append(rev_com[aa3:aa3+3])\n",
    "    \n",
    "for codon3 in orf6:\n",
    "    if len(codon3) != 3:\n",
    "        orf6.remove(codon3)\n",
    "\n",
    "for i3 in orf6:\n",
    "    if i3 in aminos.keys():\n",
    "        orf6_proteins += aminos[i3]\n",
    "        \n",
    "for i in range(len(orf6_proteins)):\n",
    "    for j in range(len(orf6_proteins)):\n",
    "        if orf6_proteins[i] == 'M':\n",
    "            if orf6_proteins[j] == '*':\n",
    "                coding_seq.append(orf6_proteins[i:j:])\n",
    "\n",
    "######### Coding Region ############\n",
    "for i in coding_seq:\n",
    "    if '*' in i:\n",
    "        coding_seq.remove(i) \n",
    "\n",
    "coding_region = []\n",
    "final_region = []\n",
    "\n",
    "for i in coding_seq:\n",
    "    if i.startswith('M') and '*' not in i:\n",
    "        coding_region.append(i)\n",
    "\n",
    "for i in coding_region:\n",
    "    if i not in final_region:\n",
    "        final_region.append(i)\n",
    "        \n",
    "for i in final_region:\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "558b3f1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def GetSeq(file):\n",
    "    \n",
    "#     file = 'rosalind_orf.txt'\n",
    "\n",
    "    with open(file, 'r') as dna:  #open fasta file\n",
    "        data = dna.readlines()\n",
    "\n",
    "    seq = ''\n",
    "\n",
    "    for row in data[1:]:      #iterate through file after header and strip newline characters\n",
    "        fasta = (row.strip('\\n'))\n",
    "        seq = seq + fasta     #join to empty list\n",
    "\n",
    "    dna.close()  #close file\n",
    "    \n",
    "    return seq\n",
    "\n",
    "def OpenReadingFrames(seq):\n",
    "    \n",
    "    aminos = {'AAA': 'K', 'AAC': 'N', 'AAG': 'K', 'AAT': 'N', 'ACA': 'T', 'ACC': 'T', 'ACG': 'T', 'ACT': 'T', 'AGA': 'R', 'AGC': 'S', 'AGG': 'R', 'AGT': 'S', 'ATA': 'I', 'ATC': 'I', 'ATG': 'M', 'ATT': 'I', 'CAA': 'Q', 'CAC': 'H', 'CAG': 'Q', 'CAT': 'H', 'CCA': 'P', 'CCC': 'P', 'CCG': 'P', 'CCT': 'P', 'CGA': 'R', 'CGC': 'R', 'CGG': 'R', 'CGT': 'R', 'CTA': 'L', 'CTC': 'L', 'CTG': 'L', 'CTT': 'L', 'GAA': 'E', 'GAC': 'D', 'GAG': 'E', 'GAT': 'D', 'GCA': 'A', 'GCC': 'A', 'GCG': 'A', 'GCT': 'A', 'GGA': 'G', 'GGC': 'G', 'GGG': 'G', 'GGT': 'G', 'GTA': 'V', 'GTC': 'V', 'GTG': 'V', 'GTT': 'V', 'TAA': '*', 'TAC': 'Y', 'TAG': '*', 'TAT': 'Y', 'TCA': 'S', 'TCC': 'S', 'TCG': 'S', 'TCT': 'S', 'TGA': '*', 'TGC': 'C', 'TGG': 'W', 'TGT': 'C', 'TTA': 'L', 'TTC': 'F', 'TTG': 'L', 'TTT': 'F'}\n",
    "\n",
    "    orf1 = []\n",
    "    orf1_proteins = ''\n",
    "\n",
    "\n",
    "    for aa in range(0,len(seq),3):   #iterate through dna seq and append every three letters to new list as codons\n",
    "        orf1.append(seq[aa:aa+3])\n",
    "\n",
    "    for i in orf1:     #iterate through list of codons and store its amino acid symbol to empty string, making protein sequence\n",
    "        if i in aminos.keys():\n",
    "            orf1_proteins += aminos[i]\n",
    "\n",
    "    ####### coding sequence ##########\n",
    "    coding_seq = []\n",
    "\n",
    "    for i in range(len(orf1_proteins)):\n",
    "        for j in range(len(orf1_proteins)):\n",
    "            if orf1_proteins[i] == 'M':\n",
    "                if orf1_proteins[j] == '*':\n",
    "                    coding_seq.append(orf1_proteins[i:j:])\n",
    "\n",
    "    ######### coding sequence ##########\n",
    "\n",
    "    orf2 = []\n",
    "    orf2_proteins = ''\n",
    "\n",
    "\n",
    "    for aa2 in range(1,len(seq),3):   #Repeat methodology in next ORF, but begin iterating through dna seq on the second character\n",
    "        orf2.append(seq[aa2:aa2+3])\n",
    "\n",
    "    for codon in orf2:        #remove item in list if it is not a full three letter codon\n",
    "        if len(codon) != 3:\n",
    "            orf2.remove(codon)\n",
    "\n",
    "    for i2 in orf2:     #translate to proteins and store in empty string\n",
    "        if i2 in aminos.keys():\n",
    "            orf2_proteins += aminos[i2]\n",
    "\n",
    "    for i in range(len(orf2_proteins)):\n",
    "        for j in range(len(orf2_proteins)):\n",
    "            if orf2_proteins[i] == 'M':\n",
    "                if orf2_proteins[j] == '*':\n",
    "                    coding_seq.append(orf2_proteins[i:j:])\n",
    "\n",
    "    ##### ORF3 #########\n",
    "\n",
    "    orf3 = []\n",
    "    orf3_proteins = ''\n",
    "\n",
    "\n",
    "    for aa3 in range(2,len(seq),3):     #same methodology as above, but again begin iterating one character down.\n",
    "        orf3.append(seq[aa3:aa3+3])\n",
    "\n",
    "    for codon3 in orf3:\n",
    "        if len(codon3) != 3:\n",
    "            orf3.remove(codon3)\n",
    "\n",
    "    for i3 in orf3:\n",
    "        if i3 in aminos.keys():\n",
    "            orf3_proteins += aminos[i3]\n",
    "\n",
    "    for i in range(len(orf3_proteins)):\n",
    "        for j in range(len(orf3_proteins)):\n",
    "            if orf3_proteins[i] == 'M':\n",
    "                if orf3_proteins[j] == '*':\n",
    "                    coding_seq.append(orf3_proteins[i:j:])\n",
    "                    \n",
    "    return coding_seq\n",
    "                \n",
    "####### REVERSE COMPLEMENT #########\n",
    "\n",
    "def ReverseComplement(seq):\n",
    "    \n",
    "    rna = ''\n",
    "\n",
    "    for nt in seq:\n",
    "        if nt == 'A':\n",
    "            rna += 'T'\n",
    "\n",
    "        elif nt == 'T':\n",
    "            rna += 'A'\n",
    "\n",
    "        elif nt == 'G':\n",
    "            rna += 'C'\n",
    "\n",
    "        elif nt == 'C':\n",
    "            rna += 'G'\n",
    "\n",
    "    rev_com = (rna[::-1])\n",
    "    \n",
    "    return rev_com\n",
    "\n",
    "def IdentifyCodingRegions(coding_seq):\n",
    "    \n",
    "    ######### Coding Region ############\n",
    "    for i in coding_seq:\n",
    "        if '*' in i:\n",
    "            coding_seq.remove(i) \n",
    "\n",
    "    coding_region = []\n",
    "    final_region = []\n",
    "\n",
    "    for i in coding_seq:\n",
    "        if i.startswith('M') and '*' not in i:\n",
    "            coding_region.append(i)\n",
    "\n",
    "    for i in coding_region:\n",
    "        if i not in final_region:\n",
    "            final_region.append(i)\n",
    "\n",
    "#     for i in final_region:\n",
    "#         print(i)\n",
    "        \n",
    "    return final_region\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8e21ba99",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['MSLLNSTKWETEGALVALERGKLPSTDTY',\n",
       " 'MR',\n",
       " 'MGAMKGSDCDR',\n",
       " 'MKGSDCDR',\n",
       " 'MCTDSSLCT',\n",
       " 'MEALGQAYELLRVGSCRVGGLSLFADCRVFGGSEDSSDKFCELSSVRMGRFKVA',\n",
       " 'MGRFKVA',\n",
       " 'MPRSQVPRTLAMFNPHESMRTRAFEI',\n",
       " 'MFNPHESMRTRAFEI',\n",
       " 'MRTRAFEI',\n",
       " 'MGNGRCSSCPGAWQVTIDGHLLA',\n",
       " 'MSSAYTQCAPIRVSAPSEVRSTKIPRSGPQIR',\n",
       " 'MNYCGLVLVG',\n",
       " 'MLTGASSASVVGSPYLRATTGYLSAPYLARCRDSNRCTLSICRRHEGQ',\n",
       " 'MSAPRLEWTVASQIPGLDFKCTGTHRFVRVKHC',\n",
       " 'MINLCGPPRLWVLRRQYRC',\n",
       " 'MVQASRCPSMVTCHAPGQLEHLPFPILWSLVGSCQRLG',\n",
       " 'MVTCHAPGQLEHLPFPILWSLVGSCQRLG',\n",
       " 'MFQAFLIV',\n",
       " 'MPGS',\n",
       " 'MRTLESSQNLSLESSDPPNTRQSANKLSPPTLQEPTRSSS',\n",
       " 'MRNYERLTNTLPHMS',\n",
       " 'MS',\n",
       " 'MQKT',\n",
       " 'MHGYA']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq = GetSeq('rosalind_orf.txt')\n",
    "rev_com = ReverseComplement(seq)\n",
    "seqORF = OpenReadingFrames(seq)\n",
    "reverse_seqORF = OpenReadingFrames(rev_com)\n",
    "\n",
    "totalORFS = seqORF + reverse_seqORF\n",
    "\n",
    "final = IdentifyCodingRegions(totalORFS)\n",
    "\n",
    "final"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "441a89e5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'MLLGSFRLIPKETLIQVAGSSPCNLS'"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq = 'AGCCATGTAGCTAACTCAGGTTACATGGGGATGACCCCGCGACTTGGATTAGAGTCTCTTTTGGAATAAGCCTGAATGATCCGAGTAGCATCTCAG'\n",
    "rev_com = ReverseComplement(seq)\n",
    "seqORF = OpenReadingFrames(seq)\n",
    "reverse_seqORF = OpenReadingFrames(rev_com)\n",
    "\n",
    "totalORFS = seqORF + reverse_seqORF\n",
    "\n",
    "final = IdentifyCodingRegions(totalORFS)\n",
    "\n",
    "longest = ''\n",
    "\n",
    "for i in final:\n",
    "    if len(i) > len(longest):\n",
    "        longest = i\n",
    "        \n",
    "longest"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
